docs/staticConfiguration.md
# Static Configuration Guide

-   [Overview](#overview)
-   [Installation](#installation)
-   [Configuration](#configuration)
-   [Usage](#usage)

# Overview

This guide will show the process of using the IoT Agent with a static configuration. In this use case, the owner of the
devices, before connecting each device, will create a static configuration for each device type. When a device is
registered, it will be assigned to one of the configured types, from where it will obtain all its configuration data.

This guide will use a Lightweight M2M client to simulate the interaction with the device. The installation and use of
this client will be explained when appropriate.

In this guide we will provide an explicit mapping for all the device attributes, using the `Robot` example given in the
[Getting Started](userGuide.md#getting-started) section. Some of them could be mapped automatically using the OMA
Registry automatic mapping (but those will be covered in other step-by-step guides).

# Installation

## Installation of the Agent

In order to install the agent, first of all, clone the GitHub repository:

```bash
git clone https://github.com/telefonicaid/lightweightm2m-iotagent.git
```

Once the repository is cloned, download the dependencies executing the following command from the root folder of the
project:

```bash
npm install
```

## Installation of the Client

In order to install the client, clone the following GitHub repository:

```bash
git clone https://github.com/telefonicaid/lwm2m-node-lib.git
```

And download the dependencies, executing, from the root folder of the project:

```bash
npm install
```

# Configuration

In order to create the new configuration, backup the default `config.js` file and edit it with your new changes.

There are two groups of data that should be changed in the configuration in order for this example to work. First of
all, in order to connect the IoT Agent to the appropriate instance of the Context Broker, you should edit the following
attributes:

-   _config.ngsi.contextBroker.host_: host IP for the ContextBroker you will be using with the IoT Agent.
-   _config.ngsi.providerUrl_: URL where your IoT Agent will be listening for ContextProvider requests. Usually this
    will be your machine's IP and the default port, but in case you are using an external context broker (or one
    deployed in a Virtual Machine) it may differ.

Once you have configured the connection information for the Context Broker, you must configure the information about the
device types you have in your system. Two attributes will have to be changed in order to accomplish this step:

-   `config.lwm2m.types`: an entry should be added to this array with information on how to map URLs to types. Edit the
    config file adding that information, as shown in the example:

```javascript
config.lwm2m = {

    [...]

    types: [
        {
            name: 'Robot',
            url: '/robots'
        }
    ]
};
```

-   `config.ngsi.types`: an entry should be added to this array with all the information about the type that is going to
    be configured. This information will include an _ad-hoc_ mapping for the attribute names to LWM2M URIs. Edit the
    config file, as shown in the example:

```javascript
config.ngsi = {

    [...]

    types: {
        'Robot': {
            service: 'Factory',
            subservice: '/robots',
            commands: [
              {
                "name": "Position",
                "type": "location"
              }
            ],
            lazy: [
              {
                "name": "Message",
                "type": "string"
              }
            ],
            attributes: [
              {
                "name": "Battery",
                "type": "number"
              }
            ],
            lwm2mResourceMapping: {
              "Battery" : {
                "objectType": 7392,
                "objectInstance": 0,
                "objectResource": 1
              },
              "Message" : {
                "objectType": 7392,
                "objectInstance": 0,
                "objectResource": 2
              },
              "Position" : {
                "objectType": 7392,
                "objectInstance": 0,
                "objectResource": 3
              }
            }
        }
    },

    [...]
};
```

You may need to change at least the log level to `DEBUG` to show information of what's going on with the execution.

# Usage

## Start the agent

In order to start the agent, from the root folder of the repository type:

```bash
bin/lwm2mAgent.js
```

This will execute the IoT Agent in the foreground, so you can see the logs of what's happening.

With the agent still open, in other terminal, open the Client, typing the following command from the client's root
folder

```bash
bin/iotagent-lwm2m-client.js
```

## Provisionings

In this case, all the information that was needed in order to work with the Agent has been already configured in the
static configuration file, so there is no need for provisioning of neither devices nor configurations.

## Using the device

In order to use the device, change to the terminal where the client has been started.

### Creation of the objects in the client

The first thing we should do before connecting to a LWM2M Server is to create the objects that the client will be
serving. The reason to perform this step before any other (specially before the registration) is that, during the
registration of the client, it will send to the server a list of all its available objects, so the server can subscribe
to those resources configured by the client as active.

From the client console, type the following commands:

```bash
LWM2M-Client> create /7392/0
```

Once the object is created, give a default value for each of the attributes:

-   Battery attribute:

```bash
LWM2M-Client> set /7392/0 1 89
```

-   Message attribute:

```bash
LWM2M-Client> set /7392/0 2 "First robot here"
```

-   Position attribute:

```bash
LWM2M-Client> set /7392/0 3 "[0, 0]"
```

### Connection to the server

Once all the objects are created in the device, connect with the server with the following command:

```bash
LWM2M-Client> connect localhost 5683 robot1 /robots
```

A few notes about this command:

-   First of all, note that the _endpoint name_ used, `robot1`, is the same we provisioned in advance with the
    provisioning request.
-   Note the URL used is `/robots`. This is the URL we configured in the `config.lwm2m.types` attribute.

The following information should be presented in the client's console:

```text
Connected:
--------------------------------
Device location: rd/2
```

This indicates that the server has accepted the connection request and assigned the `rd/2` location for the client's
requests. This exact location may change from device to device (every device has a unique location).

If you configured the server in `DEBUG` mode, check the standard output to see what happened with the client
registration.

Now you should be able to see the Entity in your Context Broker. You can do that with the following command:

```bash
(curl http://192.168.56.101:1026/v1/queryContext -s -S --header 'Content-Type: application/json' \
 --header 'Accept: application/json' --header 'fiware-service: Factory' --header 'fiware-servicepath: /robots' \
 -d @- | python -mjson.tool) <<EOF
{
    "entities": [
        {
            "type": "Robot",
            "isPattern": "false",
            "id": "robot1:Robot"
        }
    ]
}
EOF
```

Note that the headers of the request to the Context Broker should match the ones you used in the Device Provisioning.

### Updating the active attributes

In order to update the value of an attribute, issue a new `set` command, like the following:

```bash
LWM2M-Client> set /7392/0 1 67
```

This should trigger an update to the Context Broker information. A new request for the stored information in the Context
Broker should show the update.

### Reading lazy attributes

In order to read the lazy attributes, just make a query to the entity specifying the attribute you want to read, as in
the following case:

```bash
(curl http://192.168.56.101:1026/v1/queryContext -s -S --header 'Content-Type: application/json' \
 --header 'Accept: application/json' --header 'fiware-service: Factory' --header 'fiware-servicepath: /robots' \
 -d @- | python -mjson.tool) <<EOF
{
    "entities": [
        {
            "type": "Robot",
            "isPattern": "false",
            "id": "robot1:Robot"
        }
    ],
    "attributes" : [
        "Message"
    ]
}
EOF
```

### Sending a command to the device

Sending commands to a device works much the same as the updating of lazy attributes. In order to send a new command,
just update the command attribute in the Context Broker entity, with the following command:

```bash
(curl http://192.168.56.101:1026/v1/updateContext -s -S --header 'Content-Type: application/json' \
 --header 'Accept: application/json' --header 'fiware-service: Factory' --header 'fiware-servicepath: /robots' \
 -d @- | python -mjson.tool) <<EOF
{
    "contextElements": [
        {
            "type": "Robot",
            "isPattern": "false",
            "id": "robot1:Robot",
            "attributes": [
            {
                "name": "Position",
                "type": "location",
                "value": "[18,3]"
            }
            ]
        }
    ],
    "updateAction": "UPDATE"
}
EOF
```


docs/configurationProvisioning.md
# Configuration Provisioning guide

-   [Overview](#overview)
-   [Installation](#installation)
-   [Configuration](#configuration)
-   [Usage](#usage)

# Overview

This guide will show the process of using the IoT Agent with configuration provisioning. In this use case, the owner of
the devices, before connecting each one of them, provisions a device configuration that will be shared among all the
devices of the same type. The pieces of information that will be used to distinguish between devices will be the
`resource` and the API key.

This guide will use a Lightweight M2M client to simulate the interaction with the device. The installation and use of
this client will be explained when appropriate.

This guide will make use of the automatic OMA Registry mapping, with the example of the `WeatherBalloon` defined in the
[Getting Started](userGuide.md#getting-started) section. So, the attributes will be defined exclusively by their name,
and the mapping from and to a full LWM2M Mapping (like /387/3/23) will be performed by the IoT Agent.

# Installation

## Installation of the Agent

In order to install the agent, first of all, clone the GitHub repository:

```bash
git clone https://github.com/telefonicaid/lightweightm2m-iotagent.git
```

Once the repository is cloned, download the dependencies executing the following command from the root folder of the
project:

```bash
npm install
```

## Installation of the Client

In order to install the client, clone the following GitHub repository:

```bash
git clone https://github.com/telefonicaid/lwm2m-node-lib.git
```

And download the dependencies, executing, from the root folder of the project:

```bash
npm install
```

# Configuration

Most of the default `config.js` file coming with the repository should meet your needs for this guide, but there are two
attributes that you will want to tailor:

-   _config.ngsi.contextBroker.host_: host IP for the ContextBroker you will be using with the IoT Agent.
-   _config.ngsi.providerUrl_: URL where your IoT Agent will be listening for ContextProvider requests. Usually this
    will be your machine's IP and the default port, but in case you are using an external context broker (or one
    deployed in a Virtual Machine) it may differ.

You should change at least the log level to `DEBUG`, as in other levels it will not show information of what's going on
with the execution.

# Usage

## Start the agent

In order to start the agent, from the root folder of the repository type:

```bash
bin/lwm2mAgent.js
```

This will execute the IoT Agent in the foreground, so you can see the logs of what's happening.

With the agent still open, in other terminal, open the Client, typing the following command from the client's root
folder

```bash
bin/iotagent-lwm2m-client.js
```

## Provisioning the configuration

Before starting to use any device a device configuration should be created. In this step we will create a configuration
for all the devices of type 'WeatherBalloon'. To send the configuration creation request we will be using the `curl`
command that comes installed with any Unix-like OS.

This request has to be sent to the administrative port of the IoT Agent (default value 4041), not to the Lightweight M2M
port.

The following request creates the configuration group for devices with type `WeatherBalloon`:

```bash
(curl localhost:4041/iot/services -s -S --header 'Content-Type: application/json' \
  --header 'Accept: application/json' --header 'fiware-service: weather' --header 'fiware-servicepath: /balloons' \
  -d @- | python -mjson.tool) <<EOF
{
  "services": [
    {
      "resource": "/weatherBalloon",
      "apikey": "",
      "type": "WeatherBalloon",
      "commands": [],
      "lazy": [
        {
          "name": "Longitude",
          "type": "double"
        },
        {
          "name": "Latitude",
          "type": "double"
        },
        {
          "name": "Temperature Sensor",
          "type": "degrees"
        }
      ],
      "attributes": [
        {
          "name": "Power Control",
          "type": "Boolean"
        }
      ]
    }
  ]
}
EOF
```

## Using the device

In order to use the device, change to the terminal where the client has been started.

### Creation of the objects in the client

The first thing we should do before connecting to a LWM2M Server is to create the objects that the client will be
serving. The reason to perform this step before any other (specially before the registration) is that, during the
registration of the client, it will send to the server a list of all its available objects, so the server can subscribe
to those resources configured by the client as active.

From the client console, type the following commands:

```bash
LWM2M-Client> create /6/0
LWM2M-Client> create /3303/0
LWM2M-Client> create /3312/0
```

Once the object is created, give a default value for each of the attributes:

-   Longitude attribute:

```bash
LWM2M-Client> set /6/0 0 12
```

-   Longitude attribute:

```bash
LWM2M-Client> set /6/0 1 -4
```

-   Temperature attribute:

```bash
LWM2M-Client> set /3303/0 0 23
```

-   Power attribute:

```bash
LWM2M-Client> set /3312/0 0 On
```

### Connection to the server

Once all the objects are created in the device, connect with the server with the following command:

```bash
LWM2M-Client> connect localhost 5683 weather1 /weatherBalloon
```

A few notes about this command:

-   First of all, the _endpoint name_ used, `weather1`, can be whatever ID available; the type of device and its
    features will not be determined based on its Device ID, but based on the resource it is accessing.
-   The URL is `/weatherBalloon` the same one we used in the previous step, in the Configuration provisioning.

The following information should be presented in the client's console:

```text
Connected:
--------------------------------
Device location: rd/2
```

This indicates that the server has accepted the connection request and assigned the `rd/2` location for the client's
requests. This exact location may change from device to device (every device has a unique location).

If you configured the server in `DEBUG` mode, check the standard output to see what happened with the client
registration.

Now you should be able to see the Entity in your Context Broker. You can do that, querying any of the lazy attributes
with the following command:

```bash
(curl http://192.168.56.101:1026/v1/queryContext -s -S --header 'Content-Type: application/json' \
 --header 'Accept: application/json' --header 'fiware-service: weather' --header 'fiware-servicepath: /balloons' \
 -d @- | python -mjson.tool) <<EOF
{
    "entities": [
        {
            "type": "WeatherBalloon",
            "isPattern": "false",
            "id": "weather1:WeatherBalloon"
        }
    ],
    "attributes" : [
        "Latitude"
    ]
}
EOF
```

Note that the headers of the request to the Context Broker should match the ones you used in the Configuration
Provisioning.


config.js
/*
 * Copyright 2014 Telefonica Investigación y Desarrollo, S.A.U
 *
 * This file is part of fiware-iotagent-lib
 *
 * fiware-iotagent-lib is free software: you can redistribute it and/or
 * modify it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the License,
 * or (at your option) any later version.
 *
 * fiware-iotagent-lib is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public
 * License along with fiware-iotagent-lib.
 * If not, seehttp://www.gnu.org/licenses/.
 *
 * For those usages not covered by the GNU Affero General Public License
 * please contact with::[contacto@tid.es]
 */

var config = {};

config.lwm2m = {
    logLevel: 'DEBUG',
    port: 5683,
    defaultType: 'Device',
    ipProtocol: 'udp4',
    serverProtocol: 'udp4',
    /**
     * When a LWM2M client has active attributes, the IoT Agent sends an observe instruction for each one, just after the
     * client registers. This may cause cause an error when the client takes too long to start listening, as the
     * observe requests may not reach its destiny. This timeout (ms) is used to give the client the opportunity to
     * create the listener before the server sends the requests.
     */
    delayedObservationTimeout: 50,
    formats: [
        {
            name: 'application-vnd-oma-lwm2m/text',
            value: 1541
        },
        {
            name: 'application-vnd-oma-lwm2m/tlv',
            value: 1542
        },
        {
            name: 'application-vnd-oma-lwm2m/json',
            value: 1543
        },
        {
            name: 'application-vnd-oma-lwm2m/opaque',
            value: 1544
        }
    ],
    writeFormat: 'application-vnd-oma-lwm2m/text',
    types: []
};

config.ngsi = {
    logLevel: 'DEBUG',
    timestamp: true,
    contextBroker: {
        host: 'localhost',
        port: '1026'
    },
    server: {
        port: 4041
    },
    deviceRegistry: {
        //type: 'memory'
        type: 'mongodb'
    },
    mongodb: {
        host: 'localhost',
        port: '27017',
        db: 'iotagentlm2m'
        //replicaSet: ''
    },
    types: {},
    service: 'smartGondor',
    subservice: '/gardens',
    providerUrl: 'http://localhost:4041',
    deviceRegistrationDuration: 'P1Y',
    defaultType: 'Thing'
};

/**
 * Configuration for secured access to instances of the Context Broker secured with a PEP Proxy.
 * For the authentication mechanism to work, the authentication attribute in the configuration has to be fully
 * configured, and the authentication.enabled subattribute should have the value `true`.
 *
 * The Username and password should be considered as sensitive data and should not be stored in plaintext.
 * Either encrypt the config and decrypt when initializing the instance or use environment variables secured by
 * docker secrets.
 */
// config.authentication: {
//enabled: false,
/**
 * Type of the Identity Manager which is used when authenticating the IoT Agent.
 * Either 'oauth2' or 'keystone'
 */
//type: 'keystone',
/**
 * Name of the additional header passed to retrieve the identity of the IoT Agent
 */
//header: 'Authorization',
/**
 * Hostname of the Identity Manager.
 */
//host: 'localhost',
/**
 * Port of the Identity Manager.
 */
//port: '5000',
/**
 * URL of the Identity Manager - a combination of the above
 */
//url: 'localhost:5000',
/**
 * KEYSTONE ONLY: Username for the IoT Agent
 *  - Note this should not be stored in plaintext.
 */
//user: 'IOTA_AUTH_USER',
/**
 * KEYSTONE ONLY: Password for the IoT Agent
 *    - Note this should not be stored in plaintext.
 */
//password: 'IOTA_AUTH_PASSWORD',
/**
 * OAUTH2 ONLY: URL path for retrieving the token
 */
//tokenPath: '/oauth2/token',
/**
 * OAUTH2 ONLY: Flag to indicate whether or not the token needs to be periodically refreshed.
 */
//permanentToken: true,
/**
 * OAUTH2 ONLY: ClientId for the IoT Agent
 *    - Note this should not be stored in plaintext.
 */
//clientId: 'IOTA_AUTH_CLIENT_ID',
/**
 * OAUTH2 ONLY: ClientSecret for the IoT Agent
 *    - Note this should not be stored in plaintext.
 */
//clientSecret: 'IOTA_AUTH_CLIENT_SECRET'
//};

/**
 * flag indicating whether the node server will be executed in multi-core option (true) or it will be a
 * single-thread one (false).
 */
// config.multiCore = true;

module.exports = config;
